File: python/day02.py
from aoc import *
from itertools import pairwise


def is_safe(line):
    differences = [b - a for (a, b) in pairwise(line)]
    return (all( 1 <= d <=  3 for d in differences) or
            all(-1 >= d >= -3 for d in differences))

def has_single_bad_bit(line):
    for idx in range(len(line)):
        line_ = line[:idx] + line[idx+1:]
        if is_safe(line_):
            return True
    return False

def solve(filename):
    reports = parse_lines(read_input(filename), ints)
    return (sum(is_safe(line) for line in reports),
            sum(has_single_bad_bit(line) for line in reports))


solve(2)

File: elixir/day02.exs
defmodule Day02 do
  import Enum

  defp safe?(line) do
    differences =
      line
      |> zip(tl(line))
      |> map(fn {a, b} -> b - a end)

    all?(differences, &(&1 in 1..3)) or all?(differences, &(&1 in -3..-1))
  end

  defp single_bad_bit?(line) do
    line
    |> with_index
    |> any?(fn {_, i} -> line |> List.delete_at(i) |> safe?() end)
  end

  defp safe_2?(line), do: safe?(line) or single_bad_bit?(line)

  def solve(filename) do
    input =
      filename
      |> AOC.read_input()
      |> AOC.parse_lines(:ints)

    {input |> count(&safe?/1), input |> count(&safe_2?/1)}
  end
end

Day02.solve(2)
